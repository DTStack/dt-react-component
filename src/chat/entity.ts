import { immerable } from 'immer';

/**
 * 消息状态
 */
export enum MessageStatus {
    /**
     * 等待生成
     */
    PENDING,
    /**
     * 生成中
     */
    GENERATING,
    /**
     * 生成完成
     */
    DONE,
    /**
     * 生成终止
     */
    STOPPED,
}

export type Id = string;
type Timestamp = number;

export type ConversationProperties = {
    id: string;
    assistantId?: string;
    createdAt?: Timestamp;
    updatedAt?: Timestamp;
    title?: string;
    prompts?: Prompt[];
};

export type PromptProperties = {
    id: Id;
    assistantId?: string;
    createdAt?: Timestamp;
    title: string;
    messages?: Message[];
};

export type MessageProperties = {
    id: Id;
    assistantId?: string;
    creator?: string;
    createdAt?: Timestamp;
    content?: string;
    status?: MessageStatus;
};

/**
 * 新对话
 */
export abstract class Conversation {
    id: Id;
    // 后端 Id
    assistantId?: string;
    createdAt: Timestamp;
    updatedAt: Timestamp;
    title?: string;
    prompts: Prompt[];

    [immerable] = true;

    constructor(props: ConversationProperties) {
        this.id = props.id;
        this.assistantId = props.assistantId;
        this.createdAt = props.createdAt || new Date().valueOf();
        this.updatedAt = props.updatedAt || new Date().valueOf();
        this.title = props.title;
        this.prompts = props.prompts || [];
    }
}

/**
 * 一轮对话
 */
export abstract class Prompt {
    id: Id;
    // 后端 Id
    assistantId?: string;
    createdAt: Timestamp;
    title: string;
    messages: Message[];

    [immerable] = true;

    constructor(props: PromptProperties) {
        this.id = props.id;
        this.assistantId = props.assistantId;
        this.createdAt = props.createdAt || new Date().valueOf();
        this.title = props.title;
        this.messages = props.messages || [];
    }
}

/**
 * 消息实体
 */
export abstract class Message {
    id: Id;
    // 后端 Id
    assistantId?: string;
    creator?: string;
    createdAt: Timestamp;
    content: string;
    status: MessageStatus;

    [immerable] = true;

    constructor(props: MessageProperties) {
        this.id = props.id;
        this.creator = props.creator;
        this.assistantId = props.assistantId;
        this.createdAt = props.createdAt || new Date().valueOf();
        this.content = props.content ?? '';
        this.status = props.status ?? MessageStatus.PENDING;
    }
}
