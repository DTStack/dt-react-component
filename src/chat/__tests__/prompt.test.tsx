import React from 'react';
import { Components } from 'react-markdown';
import { render } from '@testing-library/react';

import { Prompt as PromptEntity } from '../entity';
import Prompt from '../prompt';
import Chat from '..';

jest.mock('remark-gfm', () => () => {});
class BasePrompt extends PromptEntity {}

function generatePrompt() {
    const prompt = new BasePrompt({
        id: '1',
        title: 'What is your name?',
        messages: [],
        createdAt: 1736479532239,
    });

    return prompt;
}

describe('Test Chat Prompt', () => {
    it('Match Snapshots', () => {
        expect(render(<Prompt />).asFragment()).toMatchSnapshot('empty');
        expect(
            render(<Prompt className="test" data={generatePrompt()} />).asFragment()
        ).toMatchSnapshot('default');
    });

    it('Should support components', () => {
        const data = generatePrompt();
        data.title += `\`inlineCode\``;
        const { getByTestId } = render(
            <Chat
                chat={{} as any}
                components={
                    {
                        code({ children }: any, _: any, extra: any) {
                            return (
                                <span data-testid="fakeCode" data-promptid={extra.promptId}>
                                    {children}
                                </span>
                            );
                        },
                    } as Components
                }
            >
                <Prompt className="test" data={data} />
            </Chat>
        );

        expect(getByTestId('fakeCode').dataset.promptid).toBe('1');
    });
});
