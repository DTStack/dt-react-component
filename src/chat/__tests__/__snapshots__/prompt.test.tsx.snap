// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Chat Prompt Match Snapshots: default 1`] = `
<DocumentFragment>
  <section
    class="dtc__prompt__container test"
  >
    <div
      class="dtc__prompt__wrapper"
    >
      <div
        class="dtc__prompt__content"
      >
        <div
          class="dtc__aigc__markdown"
        >
          <p>
            What is your name?
          </p>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Test Chat Prompt Match Snapshots: empty 1`] = `<DocumentFragment />`;

exports[`Test Chat Prompt Match Snapshots: empty title 1`] = `<DocumentFragment />`;
